lalrpop/src/build/mod.rs

```diff
@@ -266,53 +265,68 @@ fn lalrpop_files<P: AsRef<Path>>(root_dir: P) -> io::Result<Vec<PathBuf>> {
 }
 
 fn parse_and_normalize_grammar(session: &Session, file_text: &FileText) -> io::Result<r::Grammar> {
-    let grammar = match parser::parse_grammar(file_text.text()) {
-        Ok(grammar) => grammar,
+    let grammar = parser::parse_grammar(file_text.text())
+        .map_err(|error| report_parse_error(file_text, error, report_error))?;
 
-        Err(ParseError::InvalidToken { location }) => {
+    match normalize::normalize(session, grammar) {
+        Ok(grammar) => Ok(grammar),
+        Err(error) => Err(report_error(
+            file_text,
+            error.span,
+            &error.message,
+            io::Error::from(io::ErrorKind::InvalidData),
+        ))?,
+    }
+}
+
+pub fn report_parse_error<E>(
```

Comment:
This function needs a doc comment.  One because the whole callback dance is a little weird, and should be explained, but 2, the ultimate goal is for this to become a public export in lalrpop_internals, and we'd like to have it documented for that.

Reply:
Indeed. I'll add the docs.